src/ripple/basics/XRPAmount.h

```diff
@@ -0,0 +1,298 @@
+        drops_ /= rhs;
+        return *this;
+    }
+
+    XRPAmount&
+    operator%= (value_type const& rhs)
+    {
+        drops_ %= rhs;
+        return *this;
+    }
+
+    XRPAmount
+    operator- () const
+    {
+        return { -drops_ };
+    }
+
+    bool
+    operator==(XRPAmount const& other) const
+    {
+        return drops_ == other.drops_;
+    }
+
+    bool
+    operator<(XRPAmount const& other) const
+    {
+        return drops_ < other.drops_;
+    }
+
+    /** Returns true if the amount is not zero */
+    explicit
+    constexpr
+    operator bool() const noexcept
+    {
+        return drops_ != 0;
+    }
+
+    /** Return the sign of the amount */
+    constexpr
+    int
+    signum() const noexcept
+    {
+        return (drops_ < 0) ? -1 : (drops_ ? 1 : 0);
+    }
+
+    /** Returns the number of drops */
+    constexpr
+    value_type
+    drops () const
+    {
+        return drops_;
+    }
+
+    Json::Value
+    json () const
+    {
+        static_assert(std::is_signed_v<value_type>, "Expected XRPAmount to be signed");
+
+        constexpr auto min = std::numeric_limits<Json::Int>::min();
+        constexpr auto max = std::numeric_limits<Json::Int>::max();
+
+        if (drops_ < min)
+            return min;
+        if (drops_ > max)
+            return max;
+        return static_cast<Json::Int>(drops_);
+    }
+
+    /** Returns the underlying value. Code SHOULD NOT call this
+        function unless the type has been abstracted away,
+        e.g. in a templated function.
+    */
+    constexpr
+    value_type
+    value () const
+    {
+        return drops_;
+    }
+
+    friend
+    std::istream&
+    operator>> (std::istream& s, XRPAmount& val)
+    {
+        s >> val.drops_;
+        return s;
+    }
+
+};
+
+// Output XRPAmount as just the drops value.
+template<class Char, class Traits>
+std::basic_ostream<Char, Traits>&
+operator<<(std::basic_ostream<Char, Traits>& os,
+    const XRPAmount& q)
+{
+    return os << q.drops();
+}
+
+inline
+std::string
+to_string (XRPAmount const& amount)
+{
+    return std::to_string (amount.drops ());
+}
+
+inline
+constexpr
+XRPAmount
+operator*(XRPAmount::value_type lhs, XRPAmount const& rhs)
+{
+    // multiplication is commutative
+    return rhs * lhs;
+}
+
+inline
+XRPAmount
+mulRatio (
+    XRPAmount const& amt,
+    std::uint32_t num,
+    std::uint32_t den,
+    bool roundUp)
+{
+    using namespace boost::multiprecision;
+
+    if (!den)
+        Throw<std::runtime_error> ("division by zero");
+
+    int128_t const amt128 (amt.drops ());
+    auto const neg = amt.drops () < 0;
+    auto const m = amt128 * num;
+    auto r = m / den;
+    if (m % den)
+    {
+        if (!neg && roundUp)
+            r += 1;
+        if (neg && !roundUp)
+            r -= 1;
+    }
+    if (r > std::numeric_limits<XRPAmount::value_type>::max ())
+        Throw<std::overflow_error> ("XRP mulRatio overflow");
+    if (r < std::numeric_limits<XRPAmount::value_type>::min ())
```

Comment:
Strictly speaking, this is transaction breaking. However, I don't think it comes up in practice, and I do think the check should be here. I'd rather not have to put a switch here. Let me think about if we can get away without one (I think we can).

Reply:
I did a little test, and discovered that the `convert_to` function will convert an underflow to `min()`, so if I undo this change, the result will at least be a really really big negative number, as opposed to some random mapping of the underlying bits. It's also worth noting that all of these calculations deal with `XRPAmount`s, and any amount that does overflow or underflow will be dealing with more XRP than can ever exist, so it'll fail somewhere else in the process.

I can go either way.

Reply:
I agree your change is an improvement. I do want to keep it. I'm just trying to decide if we need an amendment or not.

Reply:
In my latest commit ("Review feedback #3"), I commented the change out, with a todo to determine if we need an amendment, and, more significantly, updated the unit tests to verify and demonstrate the behavior. The relevant unit tests had several holes and incorrectly documented test cases.